src/util/retry.rs

```diff
@@ -0,0 +1,253 @@
+use std::error::Error;
+use std::fmt::{Debug, Display};
+use std::future::Future;
+use std::marker::PhantomData;
+use std::time::Duration;
+
+use rand::Rng;
+
+use crate::error::VssError;
+
+/// A function that performs and retries the given operation, according to a retry policy and a set
+/// of retriable errors.
+///
+/// `retriable_errors`: Defines a list of errors on which the operation should be retried.
+/// If `retriable_errors` is `None`, operation would be retried for all errors.
+/// Errors returned from the operation must implement `PartialEq` to match items in the
+/// `retriable_errors` list. `PartialEq` should skip dynamic components of error such error message.
+pub async fn retry<R, F, Fut, T, E>(mut operation: F, retry_policy: &R) -> Result<T, E>
+where
+	R: RetryPolicy<E>,
+	F: FnMut() -> Fut,
+	Fut: Future<Output = Result<T, E>>,
+	E: Error,
+{
+	let mut attempts_made = 0;
+	let mut accumulated_delay = Duration::ZERO;
+	loop {
+		match operation().await {
+			Ok(result) => return Ok(result),
+			Err(err) => {
+				attempts_made += 1;
+				if let Some(delay) =
+					retry_policy.next_delay(&RetryContext { attempts_made, accumulated_delay, error: &err })
+				{
+					tokio::time::sleep(delay).await;
+					accumulated_delay += delay;
+				} else {
+					return Err(err);
+				}
+			}
+		}
+	}
+}
+
+/// Provides the logic for how and when to perform retries.
+pub trait RetryPolicy<E: Error>: Sized {
+	/// Returns the duration to wait before trying the next attempt.
+	/// `context` represents the context of a retry operation.
+	/// `attempts_made` is the number attempts made until now, before attempting this retry.
+	/// `accumulated_delay` is the amount of artificial delay we have already waited in between previous
+	/// attempts. Does not include the time taken to execute the operation that is being retried.
+	///
+	/// If `None` is returned then no further retry attempt is made.
+	fn next_delay(&self, context: &RetryContext<E>) -> Option<Duration>;
+
+	/// Returns a new `RetryPolicy` that respects the given maximum attempts.
+	fn with_max_attempts(self, max_attempts: u32) -> MaxAttemptsRetryPolicy<Self, E> {
+		MaxAttemptsRetryPolicy { inner_policy: self, max_attempts, phantom: PhantomData }
+	}
+	/// Returns a new `RetryPolicy` that respects the given total delay.
+	fn with_max_total_delay(self, max_total_delay: Duration) -> MaxTotalDelayRetryPolicy<Self, E> {
+		MaxTotalDelayRetryPolicy { inner_policy: self, max_total_delay, phantom: PhantomData }
+	}
+	/// Returns a new `RetryPolicy` that adds jitter(random delay) to underlying policy.
+	fn with_max_jitter(self, max_jitter: Duration) -> JitteredRetryPolicy<Self, E> {
+		JitteredRetryPolicy { inner_policy: self, max_jitter, phantom: PhantomData }
```

Comment:
I believe we can avoid `PhantomData` by moving the `E` parameter from `RetryPolicy` to `next_delay`.

Reply:
Hmm... I see. But you can make it an associated type and then it only needs to be used in the base policy. Also removes a type parameter from the decorator policies that don't care about the error.

```diff
diff --git a/src/util/retry.rs b/src/util/retry.rs
index deb2f29..abc91c0 100644
--- a/src/util/retry.rs
+++ b/src/util/retry.rs
@@ -9,7 +9,7 @@ use rand::Rng;
 /// of retriable errors.
 pub async fn retry<R, F, Fut, T, E>(mut operation: F, retry_policy: &R) -> Result<T, E>
 where
-	R: RetryPolicy<E>,
+	R: RetryPolicy<E = E>,
 	F: FnMut() -> Fut,
 	Fut: Future<Output = Result<T, E>>,
 	E: Error,
@@ -35,31 +35,34 @@ where
 }
 
 /// Provides the logic for how and when to perform retries.
-pub trait RetryPolicy<E: Error>: Sized {
+pub trait RetryPolicy: Sized {
+	/// The error type returned by the called function.
+	type E: Error;
+
 	/// Returns the duration to wait before trying the next attempt.
 	/// `context` represents the context of a retry operation.
 	///
 	/// If `None` is returned then no further retry attempt is made.
-	fn next_delay(&self, context: &RetryContext<E>) -> Option<Duration>;
+	fn next_delay(&self, context: &RetryContext<Self::E>) -> Option<Duration>;
 
 	/// Returns a new `RetryPolicy` that respects the given maximum attempts.
-	fn with_max_attempts(self, max_attempts: u32) -> MaxAttemptsRetryPolicy<Self, E> {
-		MaxAttemptsRetryPolicy { inner_policy: self, max_attempts, phantom: PhantomData }
+	fn with_max_attempts(self, max_attempts: u32) -> MaxAttemptsRetryPolicy<Self> {
+		MaxAttemptsRetryPolicy { inner_policy: self, max_attempts }
 	}
 	/// Returns a new `RetryPolicy` that respects the given total delay.
-	fn with_max_total_delay(self, max_total_delay: Duration) -> MaxTotalDelayRetryPolicy<Self, E> {
-		MaxTotalDelayRetryPolicy { inner_policy: self, max_total_delay, phantom: PhantomData }
+	fn with_max_total_delay(self, max_total_delay: Duration) -> MaxTotalDelayRetryPolicy<Self> {
+		MaxTotalDelayRetryPolicy { inner_policy: self, max_total_delay }
 	}
 	/// Returns a new `RetryPolicy` that adds jitter(random delay) to underlying policy.
-	fn with_max_jitter(self, max_jitter: Duration) -> JitteredRetryPolicy<Self, E> {
-		JitteredRetryPolicy { inner_policy: self, max_jitter, phantom: PhantomData }
+	fn with_max_jitter(self, max_jitter: Duration) -> JitteredRetryPolicy<Self> {
+		JitteredRetryPolicy { inner_policy: self, max_jitter }
 	}
 	/// Skips retrying on errors that evaluate to `true` after applying `function`.
-	fn skip_retry_on_error<F>(self, function: F) -> FilteredRetryPolicy<Self, F, E>
+	fn skip_retry_on_error<F>(self, function: F) -> FilteredRetryPolicy<Self, F, Self::E>
 	where
-		F: 'static + Fn(&E) -> bool,
+		F: 'static + Fn(&Self::E) -> bool,
 	{
-		FilteredRetryPolicy { inner_policy: self, function, phantom: PhantomData }
+		FilteredRetryPolicy { inner_policy: self, function }
 	}
 }
 
@@ -82,19 +85,20 @@ pub struct RetryContext<'a, E: Error> {
 
 /// Decorates the given `RetryPolicy` and provides a `RetryPolicy` that skips retrying errors
 /// that match the provided function.
-pub struct FilteredRetryPolicy<T: RetryPolicy<E>, F, E: Error> {
+pub struct FilteredRetryPolicy<T: RetryPolicy<E = E>, F, E: Error> {
 	inner_policy: T,
 	function: F,
-	phantom: PhantomData<E>,
 }
 
-impl<T, F, E> RetryPolicy<E> for FilteredRetryPolicy<T, F, E>
+impl<T, F, E> RetryPolicy for FilteredRetryPolicy<T, F, E>
 where
-	T: RetryPolicy<E>,
+	T: RetryPolicy<E = E>,
 	F: Fn(&E) -> bool,
 	E: Error,
 {
-	fn next_delay(&self, context: &RetryContext<E>) -> Option<Duration> {
+	type E = E;
+
+	fn next_delay(&self, context: &RetryContext<Self::E>) -> Option<Duration> {
 		if (self.function)(&context.error) {
 			None
 		} else {
@@ -105,19 +109,17 @@ where
 
 /// Decorates the given `RetryPolicy` and provides a `RetryPolicy` that respects the given maximum
 /// attempts.
-pub struct MaxAttemptsRetryPolicy<T: RetryPolicy<E>, E: Error> {
+pub struct MaxAttemptsRetryPolicy<T: RetryPolicy> {
 	/// The underlying retry policy to use.
 	inner_policy: T,
 	/// The maximum number of attempts to retry.
 	max_attempts: u32,
-	phantom: PhantomData<E>,
 }
 
-impl<T: RetryPolicy<E>, E> RetryPolicy<E> for MaxAttemptsRetryPolicy<T, E>
-where
-	E: Error,
-{
-	fn next_delay(&self, context: &RetryContext<E>) -> Option<Duration> {
+impl<T: RetryPolicy> RetryPolicy for MaxAttemptsRetryPolicy<T> {
+	type E = T::E;
+
+	fn next_delay(&self, context: &RetryContext<Self::E>) -> Option<Duration> {
 		if self.max_attempts == context.attempts_made {
 			None
 		} else {
@@ -128,19 +130,17 @@ where
 
 /// Decorates the given `RetryPolicy` and provides a `RetryPolicy` that respects the given maximum
 /// total delay.
-pub struct MaxTotalDelayRetryPolicy<T: RetryPolicy<E>, E>
-where
-	E: Error,
-{
+pub struct MaxTotalDelayRetryPolicy<T: RetryPolicy> {
 	/// The underlying retry policy to use.
 	inner_policy: T,
 	/// The maximum accumulated delay that will be allowed over all attempts.
 	max_total_delay: Duration,
-	phantom: PhantomData<E>,
 }
 
-impl<T: RetryPolicy<E>, E: Error> RetryPolicy<E> for MaxTotalDelayRetryPolicy<T, E> {
-	fn next_delay(&self, context: &RetryContext<E>) -> Option<Duration> {
+impl<T: RetryPolicy> RetryPolicy for MaxTotalDelayRetryPolicy<T> {
+	type E = T::E;
+
+	fn next_delay(&self, context: &RetryContext<Self::E>) -> Option<Duration> {
 		let next_delay = self.inner_policy.next_delay(context);
 		if let Some(next_delay) = next_delay {
 			if self.max_total_delay < context.accumulated_delay + next_delay {
@@ -154,22 +154,26 @@ impl<T: RetryPolicy<E>, E: Error> RetryPolicy<E> for MaxTotalDelayRetryPolicy<T,
 /// The exponential backoff strategy is a retry approach that doubles the delay between retries.
 /// A combined exponential backoff and jitter strategy is recommended that is ["Exponential Backoff and Jitter"](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/).
 /// This is helpful to avoid [Thundering Herd Problem](https://en.wikipedia.org/wiki/Thundering_herd_problem).
-pub struct ExponentialBackoffRetryPolicy {
+pub struct ExponentialBackoffRetryPolicy<E> {
 	/// The base delay duration for the backoff algorithm. First retry is `base_delay` after first attempt.
 	base_delay: Duration,
+
+	phantom: PhantomData<E>,
 }
 
-impl ExponentialBackoffRetryPolicy {
+impl<E: Error> ExponentialBackoffRetryPolicy<E> {
 	/// Constructs a new instance using `base_delay`.
 	///
 	/// `base_delay` is the base delay duration for the backoff algorithm. First retry is `base_delay`
 	/// after first attempt.
-	pub fn new(base_delay: Duration) -> ExponentialBackoffRetryPolicy {
-		ExponentialBackoffRetryPolicy { base_delay }
+	pub fn new(base_delay: Duration) -> Self {
+		Self { base_delay, phantom: PhantomData }
 	}
 }
 
-impl<E: Error> RetryPolicy<E> for ExponentialBackoffRetryPolicy {
+impl<E: Error> RetryPolicy for ExponentialBackoffRetryPolicy<E> {
+	type E = E;
+
 	fn next_delay(&self, context: &RetryContext<E>) -> Option<Duration> {
 		let backoff_factor = 2_u32.pow(context.attempts_made) - 1;
 		let delay = self.base_delay * backoff_factor;
@@ -179,16 +183,17 @@ impl<E: Error> RetryPolicy<E> for ExponentialBackoffRetryPolicy {
 
 /// Decorates the given `RetryPolicy` and adds jitter (random delay) to it. This can make retries
 /// more spread out and less likely to all fail at once.
-pub struct JitteredRetryPolicy<T: RetryPolicy<E>, E: Error> {
+pub struct JitteredRetryPolicy<T: RetryPolicy> {
 	/// The underlying retry policy to use.
 	inner_policy: T,
 	/// The maximum amount of random jitter to apply to the delay.
 	max_jitter: Duration,
-	phantom: PhantomData<E>,
 }
 
-impl<T: RetryPolicy<E>, E: Error> RetryPolicy<E> for JitteredRetryPolicy<T, E> {
-	fn next_delay(&self, context: &RetryContext<E>) -> Option<Duration> {
+impl<T: RetryPolicy> RetryPolicy for JitteredRetryPolicy<T> {
+	type E = T::E;
+
+	fn next_delay(&self, context: &RetryContext<Self::E>) -> Option<Duration> {
 		if let Some(base_delay) = self.inner_policy.next_delay(context) {
 			let mut rng = rand::thread_rng();
 			let jitter = Duration::from_micros(rng.gen_range(0..self.max_jitter.as_micros() as u64));
```